lib/ddq-backend-mysql.js

```diff
@@ -0,0 +1,374 @@
+        /**
+         * Queries for jobs that have stopped, and primes them to be cleaned up.
+         */
+        function heartbeatCleanup() {
+            ddqBackendInstance.restorer = null;
+
+            ddqBackendInstance.connection.query(`UPDATE ??
+                SET isProcessing = false,
+                    owner = null,
+                    requeued = false
+                WHERE DATEDIFF(NOW(), SEC_TO_TIME(?)) > heartbeatDate
+                    AND isProcessing = true;`,
+                [
+                    ddqBackendInstance.config.table,
+                    ddqBackendInstance.config.heartbeatLifetimeSeconds
+                ], (err) => {
+                    if (err) {
+                        ddqBackendInstance.emit(EMIT_ERR, err);
+                    }
+
+                    if (ddqBackendInstance.currentlyRestoring === true) {
+                        ddqBackendInstance.restorer = timers.setTimeout(heartbeatCleanup, ddqBackendInstance.config.heartbeatCleanupDelayMs);
+                    }
+                }
+            );
+        }
+
+        ddqBackendInstance.restorer = timers.setTimeout(heartbeatCleanup, ddqBackendInstance.config.heartbeatCleanupDelayMs);
+    }
+
+
+    /**
+     * Inserts a record to the database. If it's unable to do so because the
+     * record already exists, it will check directly that the entry exists
+     * and if it does, attempt to update the record.
+     *
+     * Inserting a record will repeat as many times as are defined in the
+     * config's createMessageCycleLimit.
+     *
+     * @param {Object} ddqBackendInstance
+     * @param {Object} params
+     */
+    function trySendMessage(ddqBackendInstance, params) {
+        if (cycleCounter < ddqBackendInstance.config.createMessageCycleLimit) {
+            cycleCounter += 1;
+            ddqBackendInstance.connection.query(`INSERT INTO ?? (hash, messageBase64, topics)
+                VALUES (?, ?, ?);`,
+                [
+                    ddqBackendInstance.config.table,
+                    params.hash,
+                    params.message,
+                    params.topics
+                ], (err) => {
+                    if (err && err.code === "ER_DUP_ENTRY") {
+                        writeRecord(ddqBackendInstance, params);
+                    } else if (err) {
+                        ddqBackendInstance.emit(EMIT_ERR, err);
+                    } else {
+                        params.callback();
+                    }
+                }
+            );
+        } else {
+            params.callback(new Error("Could not send message."));
+        }
+    }
+
+
+    /**
+     * Updates a record. If there isn't an error, but no records are updated, it
+     * calls the trySendMessage method. Otherwise, it will call the callback
+     * that was provided to sendMessage.
+     *
+     * @param {Object} ddqBackendInstance
+     * @param {Object} params
+     */
+    writeRecord = function (ddqBackendInstance, params) {
```

Comment:
I'd suggest renaming this.  It doesn't write a record but instead flags that a record is requeued.